### PR TITLE
[BUGFIX] Fix immediate encoding

### PR DESCRIPTION
**Pull request by krisselden**
Simplifies immediate encoding and unifies between opcode compiler and runtime stack.

Immediates are positive and handles are negative.

### CODE REVIEW

File: packages/@glimmer/util/lib/immediate.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+ *
+ * @param value - the encoded immediate value
+ */
+export function decodeImmediate(value: number): null | undefined | boolean | number {
+  if (
+    value < ImmediateConstants.MIN_IMMEDIATE ||
+    value > ImmediateConstants.MAX_IMMEDIATE ||
+    !isInt(value)
+  ) {
+    throw new Error('value is not an encoded immediate');
```
Comment (pzuraq) [inline packages/@glimmer/util/lib/immediate.ts]: Could this be guarded with `DEBUG`?

File: packages/@glimmer/util/lib/immediate.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+/**
+ * Encodes a value that can be stored directly instead of being a handle.
+ *
+ * Immediates use the positive half of 32bits
+ *
+ * @param value - the value to be encoded.
+ */
+export function encodeImmediate(value: null | undefined | boolean | number) {
+  if (typeof value === 'number') {
+    if (!isInt(value)) {
```
Comment (pzuraq) [inline packages/@glimmer/util/lib/immediate.ts]: Should these checks be wrapped in `DEBUG`?

File: packages/@glimmer/opcode-compiler/lib/debug.ts [hunk 3]
```diff
@@ -189,31 +189,12 @@ function decodeRegister(register: Register): string {
...
-          return undefined;
-      }
-    case PrimitiveType.NEGATIVE:
-    case PrimitiveType.BIG_NUM:
-      return constants.getNumber(value);
-    default:
-      throw new Error('unreachable');
+  if (primitive < 0) {
+    const handle = ~primitive;
+    if (handle < 1073741824) {
```
Comment (chadhietala) [inline packages/@glimmer/opcode-compiler/lib/debug.ts]: Probably should put this in enum type.

File: packages/@glimmer/opcode-compiler/lib/debug.ts [hunk 4]
```diff
@@ -189,31 +189,12 @@ function decodeRegister(register: Register): string {
...
+      return constants.getString(handle);
+    }
+    return constants.getNumber(handle - 1073741824);
```
Comment (chadhietala) [inline packages/@glimmer/opcode-compiler/lib/debug.ts]: Probably should put this in enum type.

File: packages/@glimmer/opcode-compiler/lib/opcode-builder/encoder.ts [hunk 5]
```diff
@@ -220,13 +217,21 @@ function constant(
...
       return operand;
     case 'primitive': {
-      let { primitive, type } = operand.value;
-      let encoded = constant(constants, primitive);
-      return sizeImmediate(constants, (encoded << 3) | type, primitive);
+      switch (operand.value.type) {
+        case PrimitiveType.STRING:
+          return ~constants.string(operand.value.primitive);
+        case PrimitiveType.NUMBER:
+          return ~(constants.number(operand.value.primitive) + 1073741824);
```
Comment (chadhietala) [inline packages/@glimmer/opcode-compiler/lib/opcode-builder/encoder.ts]: Probably should put this in enum type.

File: packages/@glimmer/runtime/lib/compiled/opcodes/vm.ts [hunk 6]
```diff
@@ -34,28 +34,17 @@ APPEND_OPCODES.add(Op.Constant, (vm, { op1: other }) => {
...
-    case PrimitiveType.NEGATIVE:
-      stack.push(vm[CONSTANTS].getNumber(value));
-      break;
-    case PrimitiveType.BIG_NUM:
-      stack.push(vm[CONSTANTS].getNumber(value));
-      break;
+  if (primitive < 0) {
+    // is handle
+    let handle = ~primitive;
+    if (handle < 1073741824) {
```
Comment (chadhietala) [inline packages/@glimmer/runtime/lib/compiled/opcodes/vm.ts]: Probably should put this in enum type.

File: packages/@glimmer/runtime/lib/compiled/opcodes/vm.ts [hunk 7]
```diff
@@ -34,28 +34,17 @@ APPEND_OPCODES.add(Op.Constant, (vm, { op1: other }) => {
...
+      stack.push(vm[CONSTANTS].getString(handle));
+    } else {
+      stack.push(vm[CONSTANTS].getNumber(handle - 1073741824));
```
Comment (chadhietala) [inline packages/@glimmer/runtime/lib/compiled/opcodes/vm.ts]: Probably should put this in enum type.

Comment (krisselden) [reply packages/@glimmer/runtime/lib/compiled/opcodes/vm.ts]: will do, this just divides the handle space in 2.  in the constants for arg operands we have 2.  on the program side there is just one array for all non directly encodable things `writeJs`,  this is why that part of the code isn't shared.

File: packages/@glimmer/runtime/lib/vm/stack.ts [hunk 8]
```diff
@@ -40,17 +41,33 @@ export class InnerStack {
...
+      case 'boolean':
+      case 'undefined':
+        this.writeRaw(pos, encodeImmediate(value));
+        break;
+      case 'object':
+        if (value === null) {
+          this.writeRaw(pos, encodeImmediate(value));
+        } else {
+          this.writeJs(pos, value);
+        }
```
Comment (krisselden) [inline packages/@glimmer/runtime/lib/vm/stack.ts]: I forgot a break here, don’t we have a lint for this?

**Pull request merged**